### PR TITLE
Use base::OnceCallback for BatAds methods

### PR DESCRIPTION
**Pull request by aseren**
Resolves https://github.com/brave/brave-browser/issues/26588

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/services/bat_ads/bat_ads_impl.h [hunk 1]
```diff
@@ -140,34 +137,6 @@ class BatAdsImpl : public mojom::BatAds,
...
-  };
-
-  static void OnMaybeServeInlineContentAd(
-      CallbackHolder<MaybeServeInlineContentAdCallback>* holder,
-      const std::string& dimensions,
-      const absl::optional<ads::InlineContentAdInfo>& ad);
-
-  static void OnMaybeServeNewTabPageAd(
-      CallbackHolder<MaybeServeNewTabPageAdCallback>* holder,
-      const absl::optional<ads::NewTabPageAdInfo>& ad);
```
Comment (tmancey) [inline components/services/bat_ads/bat_ads_impl.h:169]: `#include "third_party/abseil-cpp/absl/types/optional.h"` can be removed

File: vendor/bat-native-ads/src/bat/ads/internal/ads/inline_content_ad.cc [hunk 2]
```diff
@@ -5,6 +5,8 @@
 
 #include "bat/ads/internal/ads/inline_content_ad.h"
 
+#include <utility>
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/inline_content_ad.cc:8]: optional.h header can be removed

File: vendor/bat-native-ads/src/bat/ads/internal/ads/new_tab_page_ad.cc [hunk 3]
```diff
@@ -5,6 +5,8 @@
 
 #include "bat/ads/internal/ads/new_tab_page_ad.h"
 
+#include <utility>
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/new_tab_page_ad.cc:8]: optional.h can be removed

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/inline_content_ads/eligible_inline_content_ads_v1.cc [hunk 4]
```diff
@@ -97,41 +109,53 @@ void EligibleAdsV1::GetForChildSegments(
...
-
-        callback(/*had_opportunity*/ true, eligible_creative_ads);
-      });
+      base::BindOnce(&EligibleAdsV1::OnGetForChildSegments,
+                     base::Unretained(this), std::move(user_model), dimensions,
+                     ad_events, browsing_history, std::move(callback)));
+}
+
+void EligibleAdsV1::OnGetForChildSegments(
+    targeting::UserModelInfo user_model,
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/inline_content_ads/eligible_inline_content_ads_v1.cc:118]: clang: The parameter 'user_model' is copied for each invocation but only used as a const reference; consider making it a const reference

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/inline_content_ads/eligible_inline_content_ads_v1.h [hunk 5]
```diff
@@ -36,43 +37,77 @@ class EligibleAdsV1 final : public EligibleAdsBase {
...
+      GetEligibleAdsCallback<CreativeInlineContentAdList> callback);
+
+  void OnGetForChildSegments(
+      targeting::UserModelInfo user_model,
       const std::string& dimensions,
       const AdEventList& ad_events,
       const BrowsingHistoryList& browsing_history,
-      const GetEligibleAdsCallback<CreativeInlineContentAdList>& callback);
+      GetEligibleAdsCallback<CreativeInlineContentAdList> callback,
+      const bool success,
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/inline_content_ads/eligible_inline_content_ads_v1.h:78]: we can remove const from the declaration

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/inline_content_ads/eligible_inline_content_ads_v1.h [hunk 6]
```diff
@@ -36,43 +37,77 @@ class EligibleAdsV1 final : public EligibleAdsBase {
...
       const BrowsingHistoryList& browsing_history,
-      const GetEligibleAdsCallback<CreativeInlineContentAdList>& callback);
+      GetEligibleAdsCallback<CreativeInlineContentAdList> callback);
+
+  void OnGetForUntargeted(
+      const AdEventList& ad_events,
+      const BrowsingHistoryList& browsing_history,
+      GetEligibleAdsCallback<CreativeInlineContentAdList> callback,
+      const bool success,
+      const SegmentList& /*segments*/,
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/inline_content_ads/eligible_inline_content_ads_v1.h:109]: we can remove const from the declaration

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/inline_content_ads/eligible_inline_content_ads_v2.h [hunk 7]
```diff
@@ -36,24 +36,39 @@ class EligibleAdsV2 final : public EligibleAdsBase {
...
+      targeting::UserModelInfo user_model,
       const std::string& dimensions,
       GetEligibleAdsCallback<CreativeInlineContentAdList> callback) override;
 
  private:
+  void OnGetForUserModel(
+      targeting::UserModelInfo user_model,
+      const std::string& dimensions,
+      GetEligibleAdsCallback<CreativeInlineContentAdList> callback,
+      const bool success,
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/inline_content_ads/eligible_inline_content_ads_v2.h:48]: we can remove const from the declaration

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/inline_content_ads/eligible_inline_content_ads_v2.h [hunk 8]
```diff
@@ -36,24 +36,39 @@ class EligibleAdsV2 final : public EligibleAdsBase {
...
-      const GetEligibleAdsCallback<CreativeInlineContentAdList>& callback,
+      GetEligibleAdsCallback<CreativeInlineContentAdList> callback,
       const BrowsingHistoryList& browsing_history);
 
+  void OnGetEligibleAds(
+      const targeting::UserModelInfo& user_model,
+      const AdEventList& ad_events,
+      const BrowsingHistoryList& browsing_history,
+      GetEligibleAdsCallback<CreativeInlineContentAdList> callback,
+      const bool success,
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/inline_content_ads/eligible_inline_content_ads_v2.h:69]: we can remove const from the declaration

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/new_tab_page_ads/eligible_new_tab_page_ads_v1.h [hunk 9]
```diff
@@ -34,37 +35,67 @@ class EligibleAdsV1 final : public EligibleAdsBase {
...
   void GetForUserModel(
-      const targeting::UserModelInfo& user_model,
+      targeting::UserModelInfo user_model,
       GetEligibleAdsCallback<CreativeNewTabPageAdList> callback) override;
 
  private:
+  void OnGetForUserModel(
+      targeting::UserModelInfo user_model,
+      GetEligibleAdsCallback<CreativeNewTabPageAdList> callback,
+      const bool success,
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/new_tab_page_ads/eligible_new_tab_page_ads_v1.h:45]: we can remove const from the declaration

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/new_tab_page_ads/eligible_new_tab_page_ads_v1.h [hunk 10]
```diff
@@ -34,37 +35,67 @@ class EligibleAdsV1 final : public EligibleAdsBase {
...
+      const BrowsingHistoryList& browsing_history,
+      GetEligibleAdsCallback<CreativeNewTabPageAdList> callback);
+
+  void OnGetForChildSegments(
       const targeting::UserModelInfo& user_model,
       const AdEventList& ad_events,
       const BrowsingHistoryList& browsing_history,
-      const GetEligibleAdsCallback<CreativeNewTabPageAdList>& callback);
+      GetEligibleAdsCallback<CreativeNewTabPageAdList> callback,
+      const bool success,
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/new_tab_page_ads/eligible_new_tab_page_ads_v1.h:69]: we can remove const from the declaration

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/new_tab_page_ads/eligible_new_tab_page_ads_v1.h [hunk 11]
```diff
@@ -34,37 +35,67 @@ class EligibleAdsV1 final : public EligibleAdsBase {
...
       const AdEventList& ad_events,
       const BrowsingHistoryList& browsing_history,
-      const GetEligibleAdsCallback<CreativeNewTabPageAdList>& callback);
+      GetEligibleAdsCallback<CreativeNewTabPageAdList> callback);
+
+  void OnGetForParentSegments(
+      const AdEventList& ad_events,
+      const BrowsingHistoryList& browsing_history,
+      GetEligibleAdsCallback<CreativeNewTabPageAdList> callback,
+      const bool success,
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/new_tab_page_ads/eligible_new_tab_page_ads_v1.h:83]: we can remove const from the declaration

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/new_tab_page_ads/eligible_new_tab_page_ads_v1.h [hunk 12]
```diff
@@ -34,37 +35,67 @@ class EligibleAdsV1 final : public EligibleAdsBase {
...
       const AdEventList& ad_events,
       const BrowsingHistoryList& browsing_history,
-      const GetEligibleAdsCallback<CreativeNewTabPageAdList>& callback);
+      GetEligibleAdsCallback<CreativeNewTabPageAdList> callback);
+
+  void OnGetForUntargeted(
+      const AdEventList& ad_events,
+      const BrowsingHistoryList& browsing_history,
+      GetEligibleAdsCallback<CreativeNewTabPageAdList> callback,
+      const bool success,
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/new_tab_page_ads/eligible_new_tab_page_ads_v1.h:96]: we can remove const from the declaration

File: vendor/bat-native-ads/src/bat/ads/internal/ads/inline_content_ad.cc (hunk 2 above)
Comment (aseren) [reply vendor/bat-native-ads/src/bat/ads/internal/ads/inline_content_ad.cc:8]: fixed

File: vendor/bat-native-ads/src/bat/ads/internal/ads/new_tab_page_ad.cc (hunk 3 above)
Comment (aseren) [reply vendor/bat-native-ads/src/bat/ads/internal/ads/new_tab_page_ad.cc:8]: fixed

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/inline_content_ads/eligible_inline_content_ads_v1.cc (hunk 4 above)
Comment (aseren) [reply vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/inline_content_ads/eligible_inline_content_ads_v1.cc]: fixed

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/new_tab_page_ads/eligible_new_tab_page_ads_v1.cc [hunk 13]
```diff
@@ -31,54 +33,63 @@ EligibleAdsV1::EligibleAdsV1(
...
-    const targeting::UserModelInfo& user_model,
+    targeting::UserModelInfo user_model,
     const AdEventList& ad_events,
     const BrowsingHistoryList& browsing_history,
-    const GetEligibleAdsCallback<CreativeNewTabPageAdList>& callback) {
+    GetEligibleAdsCallback<CreativeNewTabPageAdList> callback) {
   const SegmentList segments = targeting::GetTopChildSegments(user_model);
   if (segments.empty()) {
-    GetForParentSegments(user_model, ad_events, browsing_history, callback);
+    GetForParentSegments(std::move(user_model), ad_events, browsing_history,
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/new_tab_page_ads/eligible_new_tab_page_ads_v1.cc:91]: as discussed we can remove std::move

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/inline_content_ad_serving.h [hunk 14]
```diff
@@ -47,15 +52,21 @@ class Serving final {
...
 
   void MaybeServeAd(const std::string& dimensions,
-                    const MaybeServeInlineContentAdCallback& callback);
+                    MaybeServeInlineContentAdCallback callback);
 
  private:
+  void OnGetForUserModel(const targeting::UserModelInfo& user_model,
+                         const std::string& dimensions,
+                         MaybeServeInlineContentAdCallback callback,
+                         const bool had_opportunity,
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/serving/inline_content_ad_serving.h:61]: can remove const in declaration

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/new_tab_page_ad_serving.h [hunk 15]
```diff
@@ -45,14 +50,19 @@ class Serving final {
   void AddObserver(ServingObserver* observer);
   void RemoveObserver(ServingObserver* observer);
 
-  void MaybeServeAd(const MaybeServeNewTabPageAdCallback& callback);
+  void MaybeServeAd(MaybeServeNewTabPageAdCallback callback);
 
  private:
+  void OnGetForUserModel(MaybeServeNewTabPageAdCallback callback,
+                         const targeting::UserModelInfo& user_model,
+                         const bool had_opportunity,
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/serving/new_tab_page_ad_serving.h:58]: can remove const in declaration

File: vendor/bat-native-ads/src/bat/ads/internal/creatives/notification_ads/creative_notification_ads_database_table_unittest.cc [hunk 16]
```diff
@@ -98,12 +101,15 @@ TEST_F(BatAdsCreativeNotificationAdsDatabaseTableTest,
...
-      });
+      segments,
+      base::BindOnce(
+          [](const CreativeNotificationAdList& expected_creative_ads,
+             const bool success, const SegmentList& /*segments*/,
+             const CreativeNotificationAdList& creative_ads) {
+            EXPECT_TRUE(success);
+            EXPECT_TRUE(CompareAsSets(expected_creative_ads, creative_ads));
+          },
+          std::move(expected_creative_ads)));
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/creatives/notification_ads/creative_notification_ads_database_table_unittest.cc:112]: std::move can be removed as expected_creative_ads is const

File: vendor/bat-native-ads/src/bat/ads/internal/creatives/notification_ads/creative_notification_ads_database_table_unittest.cc [hunk 17]
```diff
@@ -192,12 +198,15 @@ TEST_F(BatAdsCreativeNotificationAdsDatabaseTableTest,
...
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/creatives/notification_ads/creative_notification_ads_database_table_unittest.cc:209]: std::move can be removed as expected_creative_ads is const

File: vendor/bat-native-ads/src/bat/ads/internal/creatives/notification_ads/creative_notification_ads_database_table_unittest.cc [hunk 18]
```diff
@@ -240,12 +249,15 @@ TEST_F(BatAdsCreativeNotificationAdsDatabaseTableTest,
...
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/creatives/notification_ads/creative_notification_ads_database_table_unittest.cc:260]: std::move can be removed as expected_creative_ads is const

File: vendor/bat-native-ads/src/bat/ads/internal/creatives/notification_ads/creative_notification_ads_database_table_unittest.cc [hunk 19]
```diff
@@ -320,12 +332,15 @@ TEST_F(BatAdsCreativeNotificationAdsDatabaseTableTest,
...
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/creatives/notification_ads/creative_notification_ads_database_table_unittest.cc:343]: std::move can be removed as expected_creative_ads is const

File: vendor/bat-native-ads/src/bat/ads/internal/creatives/notification_ads/creative_notification_ads_database_table_unittest.cc [hunk 20]
```diff
@@ -414,12 +432,15 @@ TEST_F(BatAdsCreativeNotificationAdsDatabaseTableTest,
...
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/creatives/notification_ads/creative_notification_ads_database_table_unittest.cc:443]: std::move can be removed as expected_creative_ads is const

**Review by tmancey — APPROVED** [review]
LGTM++ (please see nit comments)

File: vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/notification_ads/eligible_notification_ads_v2.cc [hunk 21]
```diff
@@ -28,79 +30,96 @@ EligibleAdsV2::EligibleAdsV2(
...
+    GetEligibleAdsCallback<CreativeNotificationAdList> callback,
+    const bool success,
+    const AdEventList& ad_events) {
+  if (!success) {
+    BLOG(1, "Failed to get ad events");
+    std::move(callback).Run(/*had_opportunity*/ false, {});
+    return;
+  }
+
+  GetBrowsingHistory(user_model, ad_events, std::move(callback));
```
Comment (tmancey) [inline vendor/bat-native-ads/src/bat/ads/internal/ads/serving/eligible_ads/pipelines/notification_ads/eligible_notification_ads_v2.cc:57]: std::move(user_model)

**Review by kylehickinson — APPROVED** [review]
iOS++

**Pull request merged**